# Transfer Learning

In this notebook, we will use the bottleneck features to train a model to predict the cats. For each of the three transfer learning models, we will try some alternatives to find the best one in terms of validation accuracy.

For every model we will test three alternatives: the first one with just a 1-neuron layer to predict the final class, the second one adding a 10-neuron fully connected layer before the last layer and the third alternative adding a dropout layer (rate of 0.2) after the 10-neuron layer.

To train every model and find it's validation performance we will run 5 times and collect the average accuracy. In each of the runs, we will generate a new train and validation set, use the training set to train the model and evaluate the accuracy in the validation set. 

After all models were run, we select the best one to be the reference and find its test set accuracy, to compare with the benchmark model obtained on the prior notebook.

In [50]:
from keras.layers import Dropout, Dense
from keras.models import Sequential
from keras.callbacks import ModelCheckpoint  
from keras.backend import clear_session
from tensorflow import set_random_seed
import numpy as np
from numpy.random import seed
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
import pandas as pd

## NasNet Large

#### Importing the data

In [44]:
X = np.load('data/bottleneck_features/nasnet_bottleneck.npy')
y = np.load('data/processed_data/y.npy')

X.shape

(400, 4032)

In [45]:
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.15, random_state=42, shuffle=True, stratify=y)

#### Training the models

In [48]:
accuracy = []
epochs = 50

for i in range(5):
    # Split the data into train/validation
    X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.15, 
                                                  random_state=i + 42, shuffle=True, stratify=y_train_val)
    
    # Define the model
    model = Sequential()
    model.add(Dense(1, activation='relu', input_shape=(4032,)))
    
    # Set the random seeds
    seed(i + 42)
    set_random_seed(i + 42)
    
    # Compile and train the model
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    checkpointer = ModelCheckpoint(filepath='models/transfer_learning/weights.best.nasnet_1.hdf5', 
                               verbose=0, save_best_only=True)
    model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=epochs, 
               batch_size=32, callbacks=[checkpointer], verbose=0)
    
    # Calculate the validation accuracy
    val_predictions = (model.predict(X_val) > 0.5) * 1
    val_accuracy = accuracy_score(val_predictions, y_val)
    accuracy.append(val_accuracy)
    print('Model {}, accuracy: {}'.format(i+1, np.round(val_accuracy, 4)))
    
print('The model mean validation accuracy was: {}'.format(np.round(np.mean(accuracy),4)))

# Clear the model for the further models
clear_session()

Model 1, accuracy: 0.9216
Model 2, accuracy: 0.9608
Model 3, accuracy: 0.9412
Model 4, accuracy: 0.9608
Model 5, accuracy: 0.9804
The model mean validation accuracy was: 0.9529


In [31]:
accuracy = []
epochs = 50

for i in range(5):
    # Split the data into train/validation
    X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.15, 
                                                  random_state=i + 42, shuffle=True, stratify=y_train_val)
    
    # Define the model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(4032,)))
    model.add(Dense(1, activation='relu'))
    
    # Set the random seeds
    seed(i + 42)
    set_random_seed(i + 42)
    
    # Compile and train the model
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    checkpointer = ModelCheckpoint(filepath='models/transfer_learning/weights.best.nasnet_2.hdf5', 
                               verbose=0, save_best_only=True)
    model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=epochs, 
               batch_size=32, callbacks=[checkpointer], verbose=0)
    
    # Calculate the validation accuracy
    val_predictions = (model.predict(X_val) > 0.5) * 1
    val_accuracy = accuracy_score(val_predictions, y_val)
    accuracy.append(val_accuracy)
    print('Model {}, accuracy: {}'.format(i+1, np.round(val_accuracy, 4)))
    
print('The model mean validation accuracy was: {}'.format(np.round(np.mean(accuracy),4)))

# Clear the model for the further models
clear_session()

Model 1, accuracy: 0.5294
Model 2, accuracy: 0.9412
Model 3, accuracy: 0.9216
Model 4, accuracy: 0.9216
Model 5, accuracy: 0.9804
The model mean validation accuracy was: 0.8588


In [49]:
accuracy = []
epochs = 50

for i in range(5):
    # Split the data into train/validation
    X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.15, 
                                                  random_state=i + 42, shuffle=True, stratify=y_train_val)
    
    # Define the model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(4032,)))
    model.add(Dropout(rate=0.2))
    model.add(Dense(1, activation='relu'))
    
    # Set the random seeds
    seed(i + 42)
    set_random_seed(i + 42)
    
    # Compile and train the model
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    checkpointer = ModelCheckpoint(filepath='models/transfer_learning/weights.best.nasnet_3.hdf5', 
                               verbose=0, save_best_only=True)
    model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=epochs, 
               batch_size=32, callbacks=[checkpointer], verbose=0)
    
    # Calculate the validation accuracy
    val_predictions = (model.predict(X_val) > 0.5) * 1
    val_accuracy = accuracy_score(val_predictions, y_val)
    accuracy.append(val_accuracy)
    print('Model {}, accuracy: {}'.format(i+1, np.round(val_accuracy, 4)))
    
print('The model mean validation accuracy was: {}'.format(np.round(np.mean(accuracy),4))) 

# Clear the model for the further models
clear_session()

Model 1, accuracy: 0.902
Model 2, accuracy: 0.9608
Model 3, accuracy: 0.9216
Model 4, accuracy: 0.9804
Model 5, accuracy: 0.9804
The model mean validation accuracy was: 0.949


# Xception

#### Importing the data

In [33]:
X = np.load('data/bottleneck_features/xception_bottleneck.npy')
X.shape

(400, 1000)

In [34]:
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.15, random_state=42, shuffle=True, stratify=y)

#### Training the models

In [35]:
accuracy = []
epochs = 50

for i in range(5):
    # Split the data into train/validation
    X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.15, 
                                                  random_state=i + 42, shuffle=True, stratify=y_train_val)
    
    # Define the model
    model = Sequential()
    model.add(Dense(1, activation='relu', input_shape=(1000,)))
    
    # Set the random seeds
    seed(i + 42)
    set_random_seed(i + 42)
    
    # Compile and train the model
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    checkpointer = ModelCheckpoint(filepath='models/transfer_learning/weights.best.xception_1.hdf5', 
                               verbose=0, save_best_only=True)
    model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=epochs, 
               batch_size=32, callbacks=[checkpointer], verbose=0)
    
    # Calculate the validation accuracy
    val_predictions = (model.predict(X_val) > 0.5) * 1
    val_accuracy = accuracy_score(val_predictions, y_val)
    accuracy.append(val_accuracy)
    print('Model {}, accuracy: {}'.format(i+1, np.round(val_accuracy, 4)))
    
print('The model mean validation accuracy was: {}'.format(np.round(np.mean(accuracy),4)))

# Clear the model for the further models
clear_session()

Model 1, accuracy: 0.5098
Model 2, accuracy: 0.5098
Model 3, accuracy: 0.5098
Model 4, accuracy: 0.4902
Model 5, accuracy: 0.4902
The model mean validation accuracy was: 0.502


In [36]:
accuracy = []
epochs = 50

for i in range(5):
    # Split the data into train/validation
    X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.15, 
                                                  random_state=i + 42, shuffle=True, stratify=y_train_val)
    
    # Define the model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(1000,)))
    model.add(Dense(1, activation='relu'))
    
    # Set the random seeds
    seed(i + 42)
    set_random_seed(i + 42)
    
    # Compile and train the model
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    checkpointer = ModelCheckpoint(filepath='models/transfer_learning/weights.best.xception_2.hdf5', 
                               verbose=0, save_best_only=True)
    model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=epochs, 
               batch_size=32, callbacks=[checkpointer], verbose=0)
    
    # Calculate the validation accuracy
    val_predictions = (model.predict(X_val) > 0.5) * 1
    val_accuracy = accuracy_score(val_predictions, y_val)
    accuracy.append(val_accuracy)
    print('Model {}, accuracy: {}'.format(i+1, np.round(val_accuracy, 4)))
    
print('The model mean validation accuracy was: {}'.format(np.round(np.mean(accuracy),4)))

# Clear the model for the further models
clear_session()

Model 1, accuracy: 0.7255
Model 2, accuracy: 0.8824
Model 3, accuracy: 0.8235
Model 4, accuracy: 0.4902
Model 5, accuracy: 0.7647
The model mean validation accuracy was: 0.7373


In [37]:
accuracy = []
epochs = 50

for i in range(5):
    # Split the data into train/validation
    X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.15, 
                                                  random_state=i + 42, shuffle=True, stratify=y_train_val)
    
    # Define the model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(1000,)))
    model.add(Dropout(rate=0.2))
    model.add(Dense(1, activation='relu'))
    
    # Set the random seeds
    seed(i + 42)
    set_random_seed(i + 42)
    
    # Compile and train the model
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    checkpointer = ModelCheckpoint(filepath='models/transfer_learning/weights.best.xception_3.hdf5', 
                               verbose=0, save_best_only=True)
    model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=epochs, 
               batch_size=32, callbacks=[checkpointer], verbose=0)
    
    # Calculate the validation accuracy
    val_predictions = (model.predict(X_val) > 0.5) * 1
    val_accuracy = accuracy_score(val_predictions, y_val)
    accuracy.append(val_accuracy)
    print('Model {}, accuracy: {}'.format(i+1, np.round(val_accuracy, 4)))
    
print('The model mean validation accuracy was: {}'.format(np.round(np.mean(accuracy),4)))

# Clear the model for the further models
clear_session()

Model 1, accuracy: 0.8039
Model 2, accuracy: 0.5098
Model 3, accuracy: 0.8627
Model 4, accuracy: 0.4902
Model 5, accuracy: 0.4902
The model mean validation accuracy was: 0.6314


# Inception

#### Importing the data

In [38]:
X = np.load('data/bottleneck_features/inception_bottleneck.npy')
X.shape

(400, 1536)

In [39]:
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.15, random_state=42, shuffle=True, stratify=y)

#### Training the models

In [40]:
accuracy = []
epochs = 50

for i in range(5):
    # Split the data into train/validation
    X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.15, 
                                                  random_state=i + 42, shuffle=True, stratify=y_train_val)
    
    # Define the model
    model = Sequential()
    model.add(Dense(1, activation='relu', input_shape=(1536,)))
    
    # Set the random seeds
    seed(i + 42)
    set_random_seed(i + 42)
    
    # Compile and train the model
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    checkpointer = ModelCheckpoint(filepath='models/transfer_learning/weights.best.inception_1.hdf5', 
                               verbose=0, save_best_only=True)
    model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=epochs, 
               batch_size=32, callbacks=[checkpointer], verbose=0)
    
    # Calculate the validation accuracy
    val_predictions = (model.predict(X_val) > 0.5) * 1
    val_accuracy = accuracy_score(val_predictions, y_val)
    accuracy.append(val_accuracy)
    print('Model {}, accuracy: {}'.format(i+1, np.round(val_accuracy, 4)))
    
print('The model mean validation accuracy was: {}'.format(np.round(np.mean(accuracy),4)))

# Clear the model for the further models
clear_session()

Model 1, accuracy: 0.9216
Model 2, accuracy: 0.8824
Model 3, accuracy: 0.9412
Model 4, accuracy: 0.9608
Model 5, accuracy: 0.4902
The model mean validation accuracy was: 0.8392


In [41]:
accuracy = []
epochs = 50

for i in range(5):
    # Split the data into train/validation
    X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.15, 
                                                  random_state=i + 42, shuffle=True, stratify=y_train_val)
    
    # Define the model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(1536,)))
    model.add(Dense(1, activation='relu'))
    
    # Set the random seeds
    seed(i + 42)
    set_random_seed(i + 42)
    
    # Compile and train the model
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    checkpointer = ModelCheckpoint(filepath='models/transfer_learning/weights.best.inception_2.hdf5', 
                               verbose=0, save_best_only=True)
    model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=epochs, 
               batch_size=32, callbacks=[checkpointer], verbose=0)
    
    # Calculate the validation accuracy
    val_predictions = (model.predict(X_val) > 0.5) * 1
    val_accuracy = accuracy_score(val_predictions, y_val)
    accuracy.append(val_accuracy)
    print('Model {}, accuracy: {}'.format(i+1, np.round(val_accuracy, 4)))
    
print('The model mean validation accuracy was: {}'.format(np.round(np.mean(accuracy),4)))

# Clear the model for the further models
clear_session()

Model 1, accuracy: 0.5098
Model 2, accuracy: 0.9804
Model 3, accuracy: 0.9804
Model 4, accuracy: 0.4902
Model 5, accuracy: 0.9804
The model mean validation accuracy was: 0.7882


In [42]:
accuracy = []
epochs = 50

for i in range(5):
    # Split the data into train/validation
    X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.15, 
                                                  random_state=i + 42, shuffle=True, stratify=y_train_val)
    
    # Define the model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(1536,)))
    model.add(Dropout(rate=0.2))
    model.add(Dense(1, activation='relu'))
    
    # Set the random seeds
    seed(i + 42)
    set_random_seed(i + 42)
    
    # Compile and train the model
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    checkpointer = ModelCheckpoint(filepath='models/transfer_learning/weights.best.inception_2.hdf5', 
                               verbose=0, save_best_only=True)
    model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=epochs, 
               batch_size=32, callbacks=[checkpointer], verbose=0)
    
    # Calculate the validation accuracy
    val_predictions = (model.predict(X_val) > 0.5) * 1
    val_accuracy = accuracy_score(val_predictions, y_val)
    accuracy.append(val_accuracy)
    print('Model {}, accuracy: {}'.format(i+1, np.round(val_accuracy, 4)))
    
print('The model mean validation accuracy was: {}'.format(np.round(np.mean(accuracy),4)))  

# Clear the model for the further models
clear_session()

Model 1, accuracy: 0.9216
Model 2, accuracy: 0.9804
Model 3, accuracy: 0.9804
Model 4, accuracy: 0.4902
Model 5, accuracy: 0.5098
The model mean validation accuracy was: 0.7765


# Model comparison

Below we compare the results for the three transfer learning models (Nasnet, Xception and Inception) for each of thre three models tested. We also added a column to compare the input size for each model (size of the bottleneck features):

In [57]:
d = {'M1 Accuracy': [0.9529, 0.5020, 0.8392], 
     'M2 Accuracy': [0.8588, 0.7373, 0.7882], 
     'M3 Accuracy': [0.9400, 0.6314, 0.7765],
     'Input size': [4032, 1000, 1536]}
results = pd.DataFrame(data=d, index=['Nasnet', 'Xception', 'Inception'])
results

M1 Accuracy  M2 Accuracy  M3 Accuracy  Input size
Nasnet          0.9529       0.8588       0.9400        4032
Xception        0.5020       0.7373       0.6314        1000
Inception       0.8392       0.7882       0.7765        1536

Given this comparison, it is easy to see that the Nasnet had the best total performance in the three model alternatives tested. The main reason that I found for this is the input size of the model (4 times the Xception and 2.5 times the Inception), which is a important characteristic given that we have a small dataset, so the more information we could get from each sample the better.

Comparing the three models for Nasnet we found that the best one if the first alternative, the simplest model just containing the output layer. This model had the best accuracy probably because it avoided overfitting (given it's small size) and this shows how good the pre-trained weights in the Nasnet was for this particular problem. Given this, we now will use this first model to calculate the test set accuracy, to compare with the benchmark model trained previously (0.617).

In [66]:
X = np.load('data/bottleneck_features/nasnet_bottleneck.npy')
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.15, random_state=42, shuffle=True, stratify=y)

model = Sequential()
model.add(Dense(1, activation='relu', input_shape=(4032,)))
model.load_weights('models/transfer_learning/weights.best.nasnet_1.hdf5')

test_predictions = (model.predict(X_test) > 0.5) * 1
test_accuracy = accuracy_score(test_predictions, y_test)
print(np.round(test_accuracy,4))

0.9167


The final test accuracy was 0.9167, a great improvement from the 0.617 of the benchmark model and a great result overall considering the small dataset that we have available (just 400 images). This shows the power of transfer learning to solve new problems with small datasets.

We also can see that the test accuracy was lower than the validation accuracy (from 0.9529 to 0.9167). This is an indication of overfitting, which is expected given the small size of the data. In the next notebook we will try to mitigate this problem by using image augmentation. 